[FEATURE] Binary blob support for OGR provider, processing algorithm to extract binaries

nyalldawson opened the pull request:
Instead of converting binary fields to truncated strings, we instead store their contents as QByteArray values, allowing the original binary content to be retrieved. Update/add support also exists for binary fields (for the OGR provider).

This allows for plugins and scripts to utilise binary fields, such as extracting their contents. This PR takes advantage of this new support and adds a new algorithm for extracting binary blobs from tables. 

This algorithm allows users to extract binary fields to files. The filename is set via a data defined expression, so can be taken from a column value or a more complex expression.

nyalldawson committed:
[FEATURE] Binary blob support for OGR provider

Instead of converting binary fields to truncated strings, we instead
store their contents as QByteArray values, allowing the original binary
content to be retrieved.

This allows for plugins and scripts to utilise binary fields,
such as extracting their contents.

nyalldawson committed:
[FEATURE][processing] New algorithm for extracting binary blobs from fields

Allows users to extract binary fields to files. The filename is set via
a data defined expression, so can be taken from a column value or
a more complex expression.

nyalldawson committed:
Don't try to show CRS in map layer model for aspatial tables

nyalldawson committed:
Icon for binary fields

nyalldawson committed:
Unit tests for updating OGR binary column

nyalldawson committed:
Unit tests for extract binary algorithm

nyalldawson commented:
Future TODO: create a new edit widget for binary fields, allowing their contents to be saved to a file/loaded from a file.

nirvn commented on images/themes/default/mIconFieldBinary.svg:
```diff
@@ -0,0 +1,96 @@
+   xmlns="http://www.w3.org/2000/svg"
+   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
+   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
+   height="64"
```
@nyalldawson , should this icon be 16 x 16 pixels in size? Also, don't forget to optimize it.

nyalldawson replied:
Done

nyalldawson committed:
Fix build

luipir commented:
great :)

nyalldawson committed:
Fix build on Travis

m-kuhn commented:
Finally :tada:

m-kuhn commented on src/core/qgsogrutils.cpp:
```diff
@@ -221,6 +221,17 @@ QVariant QgsOgrUtils::getOgrFeatureAttribute( OGRFeatureH ogrFet, const QgsField
+        int size = 0;
+        const GByte *b = OGR_F_GetFieldAsBinary( ogrFet, attIndex, &size );
+        QByteArray ba = QByteArray::fromRawData( reinterpret_cast<const char *>( b ), size );
+        ba.detach();
```
Why the detach?

m-kuhn replied:
http://doc.qt.io/qt-5/qbytearray.html#fromRawData

nyalldawson replied:
I'll add a comment here. It's an odd API decision on the qt side.

nyalldawson committed:
Fix size of field icons (should be 16x16), minimize

nyalldawson committed:
Add comment on detach use

nyalldawson committed:
[FEATURE][ogr] Allow creation of binary (blob) fields for compatible filetypes

nyalldawson committed:
Fix string

nyalldawson committed:
Tweak behavior of new shapefile/gpkg browser actions

nyalldawson committed:
Unit test for creation of blob field in existing table

nyalldawson committed:
Add support for binary fields to QgsVectorFileWriter

nyalldawson committed:
[FEATURE] New form widget for binary (blob) fields

This widget is available for binary fields only (and is the default
widget used for binary fields). It offers a label showing
whether the blob field is empty or not, and if non-empty shows
the content size (in bytes/kb/etc).

A drop down menu button allows users to save the current binary
contents of the field out to a disk based file, clear the contents
of a blob field, or embed binary contents by picking a file
from their system.

nyalldawson committed:
[FEATURE] Binary (blob) field support for memory layers

Allows storage of binary blob fields in memory layers

nyalldawson committed:
Unit tests for binary editor widget

nyalldawson commented:
Now comes complete with:

- support for binary fields in vector file writer
- support for binary fields in memory layers
- support for creating new binary fields for compatible ogr data formats
- a new "binary blob" editor widget, which allows extracting the blob contents to a file on disk, and embedding a file's content to a blob field

I think this is ready for merging if someone wants to review!

nirvn commented:
@nyalldawson , what about an expression function that could return a base64 string of the blob? That'd allow nice stuff to be done (embedding html img in maptips, layout atlases, etc.)

nyalldawson commented:
@nirvn that's a good idea! Let's get this one merged first and then we can keep extending in more modular PRs

nyalldawson committed:
Dox++

nyalldawson committed:
Use messagebar for binary extraction feedback
